Document drop scopes (#514)
rust-lang/reference: src/destructors.md

# Destructors

When an [initialized]&#32;[variable] or [temporary] in Rust goes out of
[scope](#drop-scopes) its *destructor* is run, or it is *dropped*. [Assignment]
also runs the destructor of its left-hand operand, if it's initialized. If a
variable has been partially initialized, only its initialized fields are
dropped.

The destructor of a type `T` consists of

1. If `T: Drop`, calling [`<T as std::ops::Drop>::drop`]
2. Recursively running the destructor of all of its fields.
    * The fields of a [struct] or [tuple] are dropped in declaration order.
    * The fields of the active [enum variant] are dropped in declaration order.

>     * The elements of an [array] or owned [slice] are dropped from the

Author: I don't think that it's specified anywhere authoritative, but judging from the comments on that RFC I think that the RFC left it out by accident.

>     * The elements of an [array] or owned [slice] are dropped from the

Reviewer: cc @nikomatsakis -- I think this looks quite reasonable and justified; can you confirm that this was intended?

>     * The elements of an [array] or owned [slice] are dropped from the

Reviewer 2: I believe it was intended, yes. But more generally, if it matches the current behavior, I think we can confirm it, because the meta-intent of RFC #1857 was to specify that built-in types will not change their drop order.

```rust
# struct PrintOnDrop(&'static str);
# impl Drop for PrintOnDrop {
#     fn drop(&mut self) {
#         println!("drop({})", self.0);
#     }
# }
loop {
    // Tuple expression doesn't finish evaluating so operands drop in reverse order
    (
        PrintOnDrop("Outer tuple first"),
        PrintOnDrop("Outer tuple second"),
        (
            PrintOnDrop("Inner tuple first"),
            PrintOnDrop("Inner tuple second"),
            break,
        ),
        PrintOnDrop("Never created"),
    );
}
```

### Constant promotion

Promotion of a value expression to a `'static` slot occurs when the expression
could be written in a constant, borrowed, and dereferencing that borrow where
the expression was originally written, without changing the runtime behavior.
That is, the promoted expression can be evaluated at compile-time and the
resulting value does not contain [interior mutability] or [destructors] (these
properties are determined based on the value where possible, e.g. `&None`
always has the type `&'static Option<_>`, as it contains nothing disallowed).

### Temporary lifetime extension

> **Note**: The exact rules for temporary lifetime extension are subject to

> > change. This is describes the current behavior only.

Reviewer 3: ```suggestion
> change. This is describing the current behavior only.
```

# Destructors

When an [initialized]&#32;[variable] or [temporary] goes out of
[scope](#drop-scopes) its *destructor* is run, or it is *dropped*. [Assignment]
also runs the destructor of its left-hand operand, if it's initialized. If a
variable has been partially initialized, only its initialized fields are
dropped.

The destructor of a type `T` consists of:

1. If `T: Drop`, calling [`<T as std::ops::Drop>::drop`]
2. Recursively running the destructor of all of its fields.

>     * The fields of a [struct] or [tuple] are dropped in declaration order.

Reviewer 2: Nit: tuples are not declared, right? I think I would say "The fields of a tuple are dropped in order." or something

    }
}

> let mut overwritten = ShowOnDrop("drops when overwritten");

Reviewer: I'd rename `ShowOnDrop` => `PrintOnDrop` since that's what it actually does.

overwritten = ShowOnDrop("drops when scope ends");

let tuple = (ShowOnDrop("Tuple first"), ShowOnDrop("Tuple second"));

let moved;
// No destructor run on assignment.
moved = ShowOnDrop("Drops when moved");

> // drops now, but is then uninitialized

Reviewer: ```suggestion
// Drops now, but is then uninitialized.
```

moved;

> let uninitialized: ShowOnDrop;

Reviewer: ```suggestion

let uninitialized: ShowOnDrop;
```

> // Only the first element is dropped

Reviewer: ```suggestion
// Only the first element is dropped.
```

let mut partially_initialized: (ShowOnDrop, ShowOnDrop);
partially_initialized.0 = ShowOnDrop("Partial tuple first");
```

> ## Drop scopes

Reviewer: Is all of this behavior specified/decided somewhere.. is some of it documenting current rustc behavior or are all parts decided upon by e.g. the lang team or e.g. by necessity due to NLL?

Author: It's a description of the current behaviour, which is mostly observable. That said, some of it could be changed: RFC 66 proposes this, but it is still unimplemented.

Reviewer: OK; my interest is mainly in not additionally tying the language team's hand due to what we add here; if the text could be clarified to note what are stable guarantees and what are not that'd be helpful.

Reviewer 2: The documentation is "in the code", as they predate RFCs. Thanks @matthewjasper for writing it up, I too had an "in progress" write-up somewhere (part of a larger amendment to RFC 66...) but I'm happy to have someone else take a crack at it.

Some relevant links:

* [terminating scopes](https://github.com/rust-lang/rust/blob/df768c5c8fcb361c4dc94b4c776d6a78c12862e1/src/librustc_passes/region.rs#L58-L77)
* [the rules for temporary scopes and lets](https://github.com/rust-lang/rust/blob/df768c5c8fcb361c4dc94b4c776d6a78c12862e1/src/librustc_passes/region.rs#L438-L492)

Each variable or temporary is associated to a *drop scope*. When control flow
leaves a drop scope all variables associated to that scope are dropped in
reverse order of declaration (for variables) or creation (for temporaries).

> Drop scopes are determined after replacing [`for`], [`if let`] and

Reviewer: ```suggestion
Drop scopes are determined after replacing [`for`], [`if let`], and
```

[`while let`] expressions with the equivalent expressions using [`match`].
Overloaded operators are not distinguished from built-in operators and [binding
modes] are not considered.

Given a function, or closure, there are drop scopes for:

* The entire function
* Each [statement]
* Each [expression]
* Each block, including the function body

>     * [Block expressions] the expression and block scopes are the same scope.

Reviewer: This doesn't read well; separate "block expressions" and "the expression" somehow?

* Each arm of a `match` expression

Drop scopes are nested within one another as follows. When multiple scopes are
left at once, such as when returning from a function, variables are dropped
from the inside outwards.

* The entire function scope is the outer most scope.
* The function body block is contained within the scope of the entire function.
* The parent of the expression in an expression statement is the scope of the
  statement.
* The parent of the initializer of a `let` statement is the `let`
  statement's scope.
* The parent of a statement scope is the scope of the block that contains the
  statement.
* The parent of the expression for a `match` guard is the scope of the arm that

>   it's for.

Reviewer: ```suggestion
  the guard is for.
```

> * The parent of the expression for a given `match` arm is that arm's scope.

Reviewer: It isn't clear that this refers to the expr on RHS of `=>`.

> * The parent of the arm scope is the scope of the match expression that it

Reviewer: ```suggestion
* The parent of the arm scope is the scope of the `match` expression that it
```

  belongs to.
* The parent of all other scopes is the scope of the immediately enclosing
  expression.

### Scopes of function parameters

All function parameters are in the scope of the entire function body, so are

> dropped last when evaluating the function. The actual function parameter is

Reviewer: ```suggestion
dropped last when evaluating the function. Actual function parameters are
```

> dropped after any named parameters that are bound to parts of it.

Reviewer: ```suggestion
dropped after any bindings introduced in each parameter's pattern.
```

```rust
# struct ShowOnDrop(&'static str);
# impl Drop for ShowOnDrop {
#     fn drop(&mut self) {
#         println!("drop({})", self.0);
#     }
# }
// Drops the second parameter, then `y`, then the first parameter, then `x`
fn patterns_in_parameters(
    (x, _): (ShowOnDrop, ShowOnDrop),
    (_, y): (ShowOnDrop, ShowOnDrop),
) {}

// drop order is 3 2 0 1
patterns_in_parameters(
    (ShowOnDrop("0"), ShowOnDrop("1")),
    (ShowOnDrop("2"), ShowOnDrop("3")),
);
```

### Scopes of local variables

Local variables declared in a `let` statement are associated to the scope of
the block that contains the `let` statement. Local variables declared in a

> match are associated to the arm scope of the `match` arm that they are declared

Reviewer: ```suggestion
`match` expression are associated to the arm scope of the `match` arm that they are declared
```

in.

```rust
# struct ShowOnDrop(&'static str);
# impl Drop for ShowOnDrop {
#     fn drop(&mut self) {
#         println!("drop({})", self.0);
#     }
# }
let declared_first = ShowOnDrop("Dropped last in outer scope");
{
    let declared_in_block = ShowOnDrop("Dropped in inner scope");
}
let declared_last = ShowOnDrop("Dropped first in outer scope");
```

> If multiple patterns are used in the same arm for a match, then an unspecified

Reviewer: ```suggestion
If multiple patterns are used in the same arm for a `match` expression, then an unspecified
```

pattern will be used to determine the drop order.

### Temporary scopes

The *temporary scope* of an expression is the scope that is used for the
temporary variable that holds the result of that expression when used in a

> [place context], unless it is promoted to a `static`.

Reviewer: "a static" meaning?

Apart from lifetime extension, the temporary scope of an expression is the
smallest scope that contains the expression and is for one of the following:

* The entire function body.
* A statement.
* The body of a [`if`], [`while`] or [`loop`] expression.
* The `else` block of an `if` expression.

> * The condition expression of an `if` or `while` expression, or a `match` guard.

Reviewer: What about the scrutinee of `match` / `if let` / `while let` ?

Author: They're not a temporary scopes. For example in `match 1 { ref mut z => (), };` the temporary for `1` lives until the end of the statement

Reviewer: That seems like noteworthy in the reference.

        (
            ShowOnDrop("Inner tuple first"),
            ShowOnDrop("Inner tuple second"),
            break,
        ),
        ShowOnDrop("Never created"),
    );
}
```

### Constant promotion

Promotion of a value expression to a `'static` slot occurs when the expression
could be written in a constant, borrowed, and dereferencing that borrow where
the expression was originally written, without changing the runtime behavior.
That is, the promoted expression can be evaluated at compile-time and the
resulting value does not contain [interior mutability] or [destructors] (these
properties are determined based on the value where possible, e.g. `&None`
always has the type `&'static Option<_>`, as it contains nothing disallowed).

### Temporary lifetime extension

The temporary scopes for expressions in `let` statements are sometimes
*extended* to the scope of the block containing the `let` statement. This is
done when the usual temporary scope would be too small, based on certain
syntactic rules. For example:

```rust
let x = &mut 0;
// Usually a temporary would be dropped by now, but the temporary for `0` lives
// to the end of the block.
println!("{}", x);
```

> If a borrow, dereference, field or tuple indexing expression has an extended

Reviewer: ```suggestion
If a borrow, dereference, field, or tuple indexing expression has an extended
```

temporary scope then so does its operand. If an indexing expression has an
extended temporary scope then the indexed expression also has an extended
temporary scope.

#### Extending based on patterns

An *extending pattern* is either

* An [identifier pattern] that binds by reference or mutable reference.
* A [struct][struct pattern], [tuple][tuple pattern], [tuple struct][tuple

>   struct pattern] or [slice][slice pattern] pattern where at least one of the

Reviewer: ```suggestion
  struct pattern], or [slice][slice pattern] pattern where at least one of the
```

  direct subpatterns is a extending pattern.

So `ref x`, `V(ref x)` and `[ref x, y]` are all extending patterns, but `x`,
`&ref x` and `&(ref x,)` are not.

If the pattern in a `let` statement is an extending pattern then the temporary
scope of the initializer expression is extended.

#### Extending based on expressions

For a let statement with an initializer, an *extending expression* is an
expression which is one of the following:

* The initializer expression.

> * The operand of a extending [borrow expression].

Reviewer: ```suggestion
* The operand of an extending [borrow expression].
```

* The operand(s) of an extending [array][array expression], [cast][cast

>   expression], [braced struct][struct expression] or [tuple][tuple expression]

Reviewer: ```suggestion
  expression], [braced struct][struct expression], or [tuple][tuple expression]
```

  expression.
* The final expression of any extending [block expression][block expressions].

> So the borrow expressions in `&mut 0`, `(&1, &mut 2)` and `Some { 0: &mut 3 }`

Reviewer: ```suggestion
So the borrow expressions in `&mut 0`, `(&1, &mut 2)`, and `Some { 0: &mut 3 }`
```

are all extending expressions, while the borrows in `&0 + &1` and

> `Some(&mut 0)` are not.

Reviewer: `Some(&mut 0)` behaves this way because it is a function call?

Reviewer: Right; can you add a note explaining why `Some(&mut 0)` is different from `Some { 0: &mut 3 }` since it may not be so clear for anyone reading this?

The operand of any extending borrow expression has its temporary scope
extended.

#### Examples

Here are some examples where expressions have extended temporary scopes:

```rust
# fn temp() {}
# trait Use { fn use_temp(&self) -> &Self { self } }
# impl Use for () {}
// The temporary that stores the result of `temp()` lives in the same scope
// as x in these cases.
let x = &temp();
let x = &temp() as &dyn Send;
let x = (&*&temp(),);
let x = { [Some { 0: &temp(), }] };
let ref x = temp();
let ref x = *&temp();
# x;
```

Here are some examples where expressions don't have extended temporary scopes:

```rust,compile_fail
# fn temp() {}
# trait Use { fn use_temp(&self) -> &Self { self } }
# impl Use for () {}

> // The temporary that stores the result of `temp()` lives only lives until the

Reviewer: ```suggestion
// The temporary that stores the result of `temp()` only lives until the
```
